rpcs3/Emu/Cell/SPUThread.h

```diff
@@ -198,36 +198,70 @@ struct spu_channel
 	// Push performing bitwise OR with previous value, may require notification
 	bool push_or(u32 value)
 	{
-		const u64 old = data.fetch_op([value](u64& data)
+		bool pushed_to_jostling = false;
+
+		const u64 old = data.fetch_op([&](u64& data)
 		{
+			if (pushed_to_jostling)
+			{
+				return;
+			}
+
+			if (data == bit_wait)
+			{
+				// Insert the pending value in special storage for waiting SPUs, leave no time in which the channel has data
+				if (jostling_value.compare_and_swap_test(bit_wait, value))
+				{
+					data = 0;
+					pushed_to_jostling = true;
+					return;
+				}
+			}
+
 			data &= ~bit_wait;
 			data |= bit_count | value;
 		});
 
-		if (old & bit_wait)
+		if (pushed_to_jostling || old == bit_wait)
 		{
 			data.notify_one();
 		}
 
-		return (old & bit_count) == 0;
-	}
-
-	bool push_and(u32 value)
-	{
-		return (data.fetch_and(~u64{value}) & value) != 0;
+		return pushed_to_jostling || (old & bit_count) == 0;
 	}
 
 	// Push unconditionally (overwriting previous value), may require notification
 	bool push(u32 value)
 	{
-		const u64 old = data.exchange(bit_count | value);
+		bool pushed_to_jostling = false;
+
+		const u64 old = data.fetch_op([&](u64& data)
+		{
+			if (pushed_to_jostling)
+			{
```

Comment:
It looks like setting data to 0 may be abandoned here.

Reply:
Fixed.